### ISSUE TITLE
[Build]: crashing during building 

### ISSUE THREAD

**Issue by LucaTamSapienza** [openvinotoolkit/openvino#22689]
### OpenVINO Version

2023.3.0

### Operating System

Ubuntu 22.04 (LTS)

### Hardware Architecture

x86 (64 bits)

### Target Platform

Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-7
Vendor ID:               GenuineIntel
  Model name:            11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    CPU family:          6
    Model:               140
    Thread(s) per core:  2
    Core(s) per socket:  4
    Socket(s):           1
    Stepping:            1
    CPU max MHz:         4200,0000
    CPU min MHz:         400,0000
    BogoMIPS:            4838.40
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon 
                         pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2api
                         c movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow fl
                         expriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw
                          avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gf
                         ni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arch_capabilities
Virtualization features: 
  Virtualization:        VT-x
Caches (sum of all):     
  L1d:                   192 KiB (4 instances)
  L1i:                   128 KiB (4 instances)
  L2:                    5 MiB (4 instances)
  L3:                    8 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-7
Vulnerabilities:         
  Gather data sampling:  Mitigation; Microcode
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Not affected
  Retbleed:              Not affected
  Spec rstack overflow:  Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
  Srbds:                 Not affected
  Tsx async abort:       Not affected

### Build issue description

I'm encountering some problems during the OpenVINO building process. This isn't the first time that something like this has happened; I opened a similar discussion a few weeks ago (#22474). When I start the build, after some time, it stops working, and CMake gives me the following output:
```yaml
[ 63%] Linking CXX executable /home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/benchmark_app
gmake[1]: *** [CMakeFiles/Makefile2:5745: src/frontends/onnx/frontend/CMakeFiles/openvino_onnx_frontend.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 63%] Built target benchmark_app
collect2: fatal error: ld terminated with signal 9 [Killed]
compilation terminated.
gmake[2]: *** [src/plugins/intel_cpu/CMakeFiles/openvino_intel_cpu_plugin.dir/build.make:4781: /home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/libopenvino_intel_cpu_plugin.so] Error 1
gmake[2]: *** Deleting file '/home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/libopenvino_intel_cpu_plugin.so'
gmake[1]: *** [CMakeFiles/Makefile2:7022: src/plugins/intel_cpu/CMakeFiles/openvino_intel_cpu_plugin.dir/all] Error 2
[ 64%] Built target pyopenvino
gmake: *** [Makefile:166: all] Error 2
luca@luca-HP-250-G8-Notebook-PC:~/Documents/OpenSourceProject/openvino/build$
```
I just can't figure out where the problem is. I've restarted the build several times, reduced the number of parallel jobs for building, and used a virtual environment (Venv) to see if something changes, but the build consistently fails with this error."

### Build scrip or step-by-step to reproduce

I've used this configuration during the building process
```sh
cmake -DCMAKE_BUILD_TYPE=Debug -DENABLE_PYTHON=ON -DENABLE_TESTS=ON -DENABLE_FUNCTIONAL_TESTS=ON -DENABLE_DEBUG_CAPS=ON -DENABLE_CPU_DEBUG_CAPS=ON -DENABLE_NCC_STYLE=ON ..
cmake --build . --parallel 4
```

### Relevant log output

```shell
[ 67%] Building CXX object src/bindings/python/src/pyopenvino/frontend/tensorflow/CMakeFiles/py_tensorflow_frontend.dir/__/__/utils/utils.cpp.o
collect2: fatal error: ld terminated with signal 9 [Killed]
compilation terminated.
make[2]: *** [src/plugins/intel_cpu/CMakeFiles/openvino_intel_cpu_plugin.dir/build.make:4781: /home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/libopenvino_intel_cpu_plugin.so] Error 1
make[2]: *** Deleting file '/home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/libopenvino_intel_cpu_plugin.so'
make[1]: *** [CMakeFiles/Makefile2:7022: src/plugins/intel_cpu/CMakeFiles/openvino_intel_cpu_plugin.dir/all] Error 2
[ 67%] Linking CXX shared module /home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/python/openvino/frontend/tensorflow/py_tensorflow_frontend.cpython-310-x86_64-linux-gnu.so
[ 67%] Built target py_tensorflow_frontend
make: *** [Makefile:166: all] Error 2
(openvino_venv) luca@luca-HP-250-G8-Notebook-PC:~/Documents/OpenSourceProject/openvino/build$ 





[ 63%] Linking CXX executable /home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/benchmark_app
gmake[1]: *** [CMakeFiles/Makefile2:5745: src/frontends/onnx/frontend/CMakeFiles/openvino_onnx_frontend.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 63%] Built target benchmark_app
collect2: fatal error: ld terminated with signal 9 [Killed]
compilation terminated.
gmake[2]: *** [src/plugins/intel_cpu/CMakeFiles/openvino_intel_cpu_plugin.dir/build.make:4781: /home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/libopenvino_intel_cpu_plugin.so] Error 1
gmake[2]: *** Deleting file '/home/luca/Documents/OpenSourceProject/openvino/bin/intel64/Debug/libopenvino_intel_cpu_plugin.so'
gmake[1]: *** [CMakeFiles/Makefile2:7022: src/plugins/intel_cpu/CMakeFiles/openvino_intel_cpu_plugin.dir/all] Error 2
[ 64%] Built target pyopenvino
gmake: *** [Makefile:166: all] Error 2
luca@luca-HP-250-G8-Notebook-PC:~/Documents/OpenSourceProject/openvino/build$
```

### Issue submission checklist

- [X] I'm reporting an issue. It's not a question.
- [X] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.

**Comment by ilya-lavrenov** [comment]
Hi @LucaTamSapienza 
Do you need to build in Debug? It takes more memory during link / build stages.
I would recommend to build in Release

**Comment by LucaTamSapienza** [comment]
No, it's not necessary for me to build in Debug. I'll try to build it in Release, and I'll let you know. Thanks for the answer 
@ilya-lavrenov .

**Comment by vurusovs** [comment]
@LucaTamSapienza could you try to disable parallel build? It significantly affects RAM consumption

**Comment by LucaTamSapienza** [comment]
At the end, it worked. i've used this configuration for the building process
```sh
-DCMAKE_BUILD_TYPE=Release -DENABLE_PYTHON=ON -DENABLE_TESTS=ON ..
```
I restarted the pc, disabled the `functional tests`, `cpu_debug_caps`, `debug_caps` and `ncc_style` and also disable the `--parallel` process (since it already gave me some problems). I'll keep you posted on whether I encounter the same issue with my new PC.
Thanks for your help!

**Comment by LucaTamSapienza** [comment]
@ilya-lavrenov, @vurusovs 
Hello, I'm writing again on the same issue to avoid creating a new one. I recently switched my computer from Linux to Mac (ARM). I'm following the documentation regarding building process. My question is whether every time I compile, I need to enter the command  `-G "Ninja Multi-Config" -DENABLE_SYSTEM_PUGIXML=ON -DENABLE_SYSTEM_SNAPPY=ON -DENABLE_SYSTEM_PROTOBUF=ON .. ` after cmake. (Like for linux was enough to type cmake .. for example.)

Since on Linux, I might have used a different configuration for cmake, like `cmake -DENABLE_TESTS=ON .. `, for example. The guide for Mac doesn't provide information on this. If I want to compile using a custom configuration, should I proceed as I did on Linux? Later in the compilation phase, the guide specifies `cmake --build . --config Release --parallel $(sysctl -n hw.ncpu)`. Looking at `--config Release`, it makes me think that this is where I should specify the cmake configuration types I mentioned earlier, right?

Switching from working on Ubuntu to MacOS is quite challenging since I had become accustomed to it. Now, I wonder if the choice I made was a mistake.
